ui/tests/acceptance/allocation-detail-test.js

```diff
@@ -72,6 +72,111 @@ module('Acceptance | allocation detail', function(hooks) {
     assert.equal(Allocation.resourceCharts.objectAt(1).name, 'Memory', 'Second chart is Memory');
   });
 
+  test('/allocation/:id should present task lifecycles', async function(assert) {
+    const job = server.create('job', {
+      groupsCount: 1,
+      groupTaskCount: 3,
+      withGroupServices: true,
+      createAllocations: false,
+    });
+
+    const allocation = server.create('allocation', 'withTaskWithPorts', 'withAllocatedResources', {
+      clientStatus: 'running',
+      jobId: job.id,
+    });
+
+    const taskStatePhases = server.db.taskStates.where({ allocationId: allocation.id }).reduce(
+      (phases, state) => {
+        const lifecycle = server.db.tasks.findBy({ name: state.name }).Lifecycle;
+
+        if (lifecycle) {
+          if (lifecycle.Sidecar) {
+            phases.sidecars.push(state);
+            state.lifecycleString = 'Sidecar';
+          } else {
+            phases.prestarts.push(state);
+            state.lifecycleString = 'Prestart';
+          }
+        } else {
+          phases.mains.push(state);
+          state.lifecycleString = 'Main';
+        }
+
+        return phases;
+      },
+      {
+        prestarts: [],
+        sidecars: [],
+        mains: [],
+      }
+    );
+
+    taskStatePhases.prestarts = taskStatePhases.prestarts.sortBy('name');
+    taskStatePhases.sidecars = taskStatePhases.sidecars.sortBy('name');
+    taskStatePhases.mains = taskStatePhases.mains.sortBy('name');
+
+    const sortedServerStates = taskStatePhases.prestarts.concat(
+      taskStatePhases.sidecars,
+      taskStatePhases.mains
+    );
+
+    await Allocation.visit({ id: allocation.id });
+
+    assert.ok(Allocation.lifecycleChart.isPresent);
+    assert.equal(Allocation.lifecycleChart.title, 'Task Lifecycle Status');
+    assert.equal(Allocation.lifecycleChart.phases.length, 2);
+    assert.equal(Allocation.lifecycleChart.tasks.length, sortedServerStates.length);
+
+    const stateActiveIterator = state => state.state === 'running';
+    const anyPrestartsActive = taskStatePhases.prestarts.some(stateActiveIterator);
+
+    if (anyPrestartsActive) {
+      assert.ok(Allocation.lifecycleChart.phases[0].isActive);
+    } else {
+      assert.notOk(Allocation.lifecycleChart.phases[0].isActive);
+    }
```

Comment:
I can't blame you for using conditionals here given how much goes into setting up state for this test, but it would be nice to ensure we are testing all life cycle permutations.

I find these types of tests will often only fail intermittently when refactoring and that can be annoying to deal with. Contrast that with tests that fail intermittently due to bad input, which is also annoying, but harder to exhaustively test.

I'm ok with kicking this down the road though.

Reply:
Thinking about this a little more, maybe this test is ok as long as there is also comprehensive integration test coverage. Testing in that context should be easier to setup and they execute faster.

Reply:
👍 I’m definitely not a fan of much dynamism in acceptance tests; this one was much worse in that sense before I added the cycle in the factory that let me guarantee the presence of all three lifecycle types of tasks. I’ve added an integration test in 4ea527c, let me know whether you think that covers it 🤓

Reply:
Yes! Those tests definitely make me feel better about comprehensive coverage.